static_frame/test/unit/test_index_hierarchy.py

```diff
@@ -1355,7 +1308,218 @@ def test_hierarchy_relabel_a(self) -> None:
+                [[9, 9, 2],
+                 [8, 8, 1],
+                 [7, 7, 2],
+                 [6, 6, 1],
+                 [5, 5, 2],
+                 [4, 4, 1],
+                 [3, 3, 2],
+                 [2, 2, 1]])
+
+    def test_hierarchy_relabel_at_depth_3d_all_depths(self) -> None:
+
+        ih = IndexHierarchy.from_product(('I', 'II'), ('B', 'A'), (2, 1))
+
+        # Mapping
+        series_map = Series([0, True, None, '13'], index=['I', 'II', 'B', 2])
+        ih1 = ih.relabel_at_depth(series_map, depth_level=[2,0,1])
+        self.assertEqual(ih1.values.tolist(),
+                [[0, None, '13'],
+                 [0, None, 1],
+                 [0, 'A', '13'],
+                 [0, 'A', 1],
+                 [True, None, '13'],
+                 [True, None, 1],
+                 [True, 'A', '13'],
+                 [True, 'A', 1]])
+
+        # Function
+        numbers = (n for n in range(1, 100000000))
+        def func(arg: tp.Any) -> int:
+            return next(numbers)
+
+        ih2 = ih.relabel_at_depth(func, depth_level=(l for l in (1, 0, 2)))
+        self.assertEqual(ih2.values.tolist(),
+                [[1, 3, 5],
+                 [1, 3, 6],
+                 [1, 4, 7],
+                 [1, 4, 8],
+                 [2, 9, 11],
+                 [2, 9, 12],
+                 [2, 10, 13],
+                 [2, 10, 14]])
+
+        # Sequence
+        ih3 = ih.relabel_at_depth(range(8), depth_level=range(3))
+        self.assertEqual(ih3.values.tolist(),
+                [[0, 0, 0],
+                 [1, 1, 1],
+                 [2, 2, 2],
+                 [3, 3, 3],
+                 [4, 4, 4],
+                 [5, 5, 5],
+                 [6, 6, 6],
+                 [7, 7, 7]])
+
+    def test_hierarchy_relabel_at_depth_bad_input(self) -> None:
+        ih = IndexHierarchy.from_product(('I', 'II'), ('A', 'B'))
+
+        # Iterable is not long enough!
+        with self.assertRaises(ValueError):
+            ih.relabel_at_depth(range(3), depth_level=0)
+
+        # Iterable is too long!
+        with self.assertRaises(ValueError):
+            ih.relabel_at_depth(range(5), depth_level=0)
+
+        # Depth levels are not unique
+        with self.assertRaises(ValueError):
+            ih.relabel_at_depth(dict(), depth_level=[0, 0])
+
+        # Depth level is too shallow
+        with self.assertRaises(ValueError):
+            ih.relabel_at_depth(dict(), depth_level=2)
+
+        # Depth level outside range positive
+        with self.assertRaises(ValueError):
+            ih.relabel_at_depth(range(4), depth_level=3)
+
+        # Depth level outside range negative
+        with self.assertRaises(ValueError):
+            ih.relabel_at_depth(range(4), depth_level=-1)
+
+        # No depth levels!
+        with self.assertRaises(ValueError):
+            ih.relabel_at_depth(lambda:None, depth_level=[])
+
+    def test_hierarchy_relabel_at_depth_properties(self) -> None:
```

Comment:
This is a great test. Maybe just make another version of the same with an `IndexHierarchyGO` to compare these assumptions in that context.